.github/actions/sonar/action.yml

```diff
@@ -0,0 +1,269 @@
+    required: true
+  workdir:
+    description: 'Working directory for the action'
+    required: true
+  docker_compose_file:
+    description: 'Path to the Docker Compose file'
+    required: true
+  github_token:
+    description: 'GitHub token for commenting on the PR'
+    required: true
+
+runs:
+  using: composite
+  steps:
+    # Cache the main branch project for future PR analysis
+    - name: 'Cache project on main branch'
+      if: github.ref == 'refs/heads/main'
+      shell: bash
+      run: |
+        mkdir -p ./${{ inputs.sonar_project_key }}-cached
+        cp -r ${{ inputs.workdir }}/. ./${{ inputs.sonar_project_key }}-cached/
+
+    - name: 'Store cache (main branch)'
+      if: github.ref == 'refs/heads/main'
+      uses: actions/cache@v4
+      with:
+        path: ./${{ inputs.sonar_project_key }}-cached
+        key: cache-${{ inputs.sonar_project_key }}-${{ github.sha }}
+
+    # Restore the cache for PR analysis
+    - name: 'Restore project cache for PR analysis'
+      if: github.event_name == 'pull_request'
+      id: restore_cache
+      uses: actions/cache@v4
+      with:
+        path: ./${{ inputs.sonar_project_key }}-cached
+        key: cache-${{ inputs.sonar_project_key }}-${{ github.event.pull_request.base.sha }}
+        restore-keys: |
+          cache-${{ inputs.sonar_project_key }}-
+
+    # Install required tools
+    - name: 'Install jq for JSON parsing'
+      if: github.event_name == 'pull_request' && steps.restore_cache.outputs.cache-hit == 'true'
+      shell: bash
+      run: sudo apt-get install -y jq
+
+    # Start SonarQube server using Docker Compose
+    - name: 'Start SonarQube server'
+      if: github.event_name == 'pull_request' && steps.restore_cache.outputs.cache-hit == 'true'
+      shell: bash
+      run: |
+        docker compose -f ${{ inputs.docker_compose_file }} up -d
+        timeout 120s bash -c 'until curl -s http://localhost:9000/api/system/status | grep -q "\"status\":\"UP\""; do sleep 5; done' || (echo "SonarQube failed to start in time" && exit 1)
+
+    # Create SonarQube project
+    - name: 'Create SonarQube project'
+      shell: bash
+      run: |
+        curl -s -u admin:admin -X POST "http://localhost:9000/api/projects/create?name=${{ inputs.sonar_project_key }}&project=${{ inputs.sonar_project_key }}" || true
+
+    # Set up a temporary Git repository with files from the main branch cache
+    - name: 'Prepare repository with main branch files'
+      if: github.event_name == 'pull_request' && steps.restore_cache.outputs.cache-hit == 'true'
+      shell: bash
+      run: |
+        mkdir -p ./sonarqube_analysis_repo
+        cd ./sonarqube_analysis_repo
+
+        cp -r ../${{ inputs.sonar_project_key }}-cached/. .
+        rm -rf .git
+
+        git init
+        git config user.name "GitHub Action"
+        git config user.email "action@github.com"
+
+        git add -A
+        git commit -m "Initial commit from main branch"
```

Comment:
I initially struggled for many hours to get SonarQube to identify changes correctly, so I decided to create a new repository to avoid any uncertainties. However, I’ve now tested using the cached Git from the application, and it works fine (as it obviously should).